File: lib/py/myrpc/transport/MemoryTransport.py
import io

from myrpc.Common import MessageTruncatedException
from myrpc.transport.TransportBase import TransportState, TransportBase

class MemoryTransport(TransportBase):
    """Provide memory-buffered transport."""

    def __init__(self, buf = None):
        """Initialize transport.

        If rbuf is specified, then its content is copied to the read memory buffer.
        """

        super().__init__()

        self._rf = io.BytesIO(buf)
        self._wf = io.BytesIO()

    def set_state(self, state):
        if state == TransportState.READ_END:
            self._rf.truncate(0)
        elif state == TransportState.WRITE_BEGIN:
            self._wf.truncate(0)

    def read(self, count):
        buf = self._rf.read(count)
        if len(buf) < count:
            raise MessageTruncatedException()

        return buf

    def write(self, buf):
        self._wf.write(buf)

    def get_value(self):
        """Return bytes containing the entire contents of the write memory buffer."""

        buf = self._wf.getvalue()

        return buf

File: runtime/js/transport/MemoryTransport.js
// MODULE_DEP ../Common
// MODULE_DEP ./TransportBase
// MODULE_DEP ../util/BufferIO
// MODULE_EXPORT

// FIXME: set_oncontinue present in interface (not needed).

myrpc.transport.MemoryTransport = function(buf)
{
    myrpc.transport.TransportBase.call(this);

    this._rbio = new myrpc.util.BufferIO(buf);
    this._wbio = new myrpc.util.BufferIO();
};

myrpc.transport.MemoryTransport.prototype = Object.create(myrpc.transport.TransportBase.prototype);
myrpc.transport.MemoryTransport.prototype.constructor = myrpc.transport.MemoryTransport;

myrpc.transport.MemoryTransport.prototype.set_state = function(state)
{
    switch (state) {
    case myrpc.transport.TransportState.READ_END:
	this._rbio.set_buffer();
	break;

    case myrpc.transport.TransportState.WRITE_BEGIN:
	this._wbio.set_buffer();
	break;
    }
};

myrpc.transport.MemoryTransport.prototype.read = function(count)
{
    var buf = this._rbio.read(count);

    if (count > buf.length)
	throw new myrpc.common.MessageTruncatedException();

    return buf;
};

myrpc.transport.MemoryTransport.prototype.write = function(buf)
{
    this._wbio.write(buf);
};

myrpc.transport.MemoryTransport.prototype.get_value = function()
{
    var buf = this._wbio.get_buffer();

    return buf;
};
